tests/unit-test.el

```diff
@@ -141,31 +141,132 @@
 
 (describe "cljr--unresolved-alias-ref"
   (it "returns unresolved alias reference"
-    (expect (cljr--with-clojure-temp-file "foo.clj"
-              (insert "(ns foo)")
-              (cljr--unresolved-alias-ref "ns-name"))
-            :to-equal "ns-name"))
+    (cljr--with-clojure-temp-file "foo.clj"
+      (insert "(ns foo)")
+      (expect (cljr--unresolved-alias-ref "ns-name")
+              :to-equal "ns-name")))
 
   (it "returns nil for resolved alias"
-    (expect (cljr--with-clojure-temp-file "foo.clj"
-              (insert "(ns foo (:require [user.ns-name :as ns-name]))")
-              (cljr--unresolved-alias-ref "ns-name"))
-            :to-be nil))
+    (cljr--with-clojure-temp-file "foo.clj"
+      (insert "(ns foo (:require [user.ns-name :as ns-name]))")
+      (expect (cljr--unresolved-alias-ref "ns-name")
+              :to-be nil)))
 
   (it "returns nil for js alias in cljs file."
-    (expect (cljr--with-clojure-temp-file "foo.cljs"
-              (insert "(ns foo)")
-              (cljr--unresolved-alias-ref "js"))
-            :to-be nil))
+    (cljr--with-clojure-temp-file "foo.cljs"
+      (insert "(ns foo)")
+      (expect (cljr--unresolved-alias-ref "js")
+              :to-be nil)))
 
   (it "returns js for js alias in clj file."
-    (expect (cljr--with-clojure-temp-file "foo.clj"
-              (insert "(ns foo)")
-              (cljr--unresolved-alias-ref "js"))
-            :to-equal "js"))
+    (cljr--with-clojure-temp-file "foo.clj"
+      (insert "(ns foo)")
+      (expect (cljr--unresolved-alias-ref "js")
+              :to-equal "js")))
 
   (it "returns js for js alias in cljc file."
-    (expect (cljr--with-clojure-temp-file "foo.cljc"
-              (insert "(ns foo)")
-              (cljr--unresolved-alias-ref "js"))
-            :to-equal "js")))
+    (cljr--with-clojure-temp-file "foo.cljc"
+      (insert "(ns foo)")
+      (expect (cljr--unresolved-alias-ref "js")
+              :to-equal "js"))))
+
+(defmacro with-point-at (text &rest body)
+  (declare (indent 1))
+  `(progn (insert ,text)
+          (goto-char (point-min))
+          (re-search-forward "|")
+          (delete-char -1)
+          ,@body))
+
+(describe "cljr--language-context-at-point"
+  (it "identifies a clj file"
+    (cljr--with-clojure-temp-file "foo.clj"
+      (insert "(ns foo)")
+      (expect (cljr--language-context-at-point)
+              :to-equal '("clj" nil))))
+
+  (it "identifies a cljs file"
+    (cljr--with-clojure-temp-file "foo.cljs"
+      (insert "(ns foo)")
+      (expect (cljr--language-context-at-point)
+              :to-equal '("cljs" nil))))
+
+  (it "identifies a cljc file"
+    (cljr--with-clojure-temp-file "foo.cljc"
+      (insert "(ns foo)")
+      (expect (cljr--language-context-at-point)
+              :to-equal '("cljc" nil))))
+
+  (it "identifies a cljc file with a cljs context"
+    (cljr--with-clojure-temp-file "foo.cljc"
+      (with-point-at "(ns foo) #?(:cljs (Math/log|))"
```

Comment:
We could exercise a few more cases:

* `?(:clj 1 :cljs (Math/log|)`
  * two branches present
* `?(:cljs (Math/log| :clj 1)`
  * two branches present, reverse order
* `?(|)`
  * incomplete reader conditional

Reply:
I don't think the current method of parsing will handle all of that. I cribbed something together to support a few cases using the existing reader conditional code. However, this should also support, :cljr, :bb, and :default and the incomplete reader conditional, and all of those cause emacs to lockup in an infinite loop as the existing `cljr--point-in-reader-conditional-branch-p` method expects an exact match or loops forever. I'm not sure how critical this is just yet though, as from my understanding the middleware is still ignoring this field.

I've add the test cases here and mark the ones that are failing as pending but I think it may out of scope to address this in this PR as it's fixes for existing code. This should definitely be addressed before the flag is enabled by default though.

Reply:
> and all of those cause emacs to lockup in an infinite loop

Well, that seems an important issue, showing that it was a good use of code review. We most certainly don't want people experiencing/reporting infinite loops.

(Especially since this PR will use that faulty code more frequently than we do now)

It would be a much welcome side quest if you could fix this (here or separately - as you wish)

Reply:
I'm going to temporarily disable the reader conditional language context field while it's locking up and disable the corresponding tests for now. That way it's not introducing a higher risk of causing emacs to block, and we can re-enable it when the middleware starts supporting that field and implement a more robust parsing solution using the existing tests. I think the API is fine here though, we can improve it as we move forward.

Reply:
Extracted into https://github.com/clojure-emacs/clj-refactor.el/issues/533